src/corecel/sys/KernelRegistry.hh

```diff
@@ -0,0 +1,126 @@
+    inline void log_launch(int num_threads);
+};
+
+//---------------------------------------------------------------------------//
+struct KernelMetadata
+{
+    std::string      name;
+    KernelAttributes attributes;
+    KernelProfiling  profiling;
+};
+
+//! Ordered identifiers for registered kernels
+using KernelId = OpaqueId<KernelMetadata>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Keep track of kernels and launches.
+ *
+ * This class has a thread-safe *insert* method because it's meant to be shared
+ * across multiple threads when running. Since accessing by ID is expected to
+ * be done well after all insertions have been completed, the \c size and \c at
+ * accessors are *not* thread safe.
+ */
+class KernelRegistry
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using value_type      = KernelMetadata;
+    using const_reference = const KernelMetadata&;
+    using key_type        = KernelId;
+    using size_type       = KernelId::size_type;
+    //!@}
+
+  public:
+    // Whether profiling metrics (launch count, max threads) are collected
+    static bool profiling();
+
+    // Construct without any data
+    KernelRegistry() = default;
+
+    // Register a kernel and return optional reference to profiling info
+    KernelProfiling* insert(const char* name, KernelAttributes&& attrs);
+
+    //! Number of kernel diagnostics available
+    size_type size() const { return kernels_.size(); }
+
+    // Access kernel data for a single kernel
+    inline const_reference at(key_type id) const;
+
+  private:
+    using UPKM = std::unique_ptr<KernelMetadata>;
+
+    std::mutex        insert_mutex_;
+    std::vector<UPKM> kernels_;
+};
+
+//---------------------------------------------------------------------------//
+// FREE FUNCTIONS
+//---------------------------------------------------------------------------//
+// Globally shared registry of kernels for end-of-program diagnostics
+KernelRegistry& kernel_registry();
+
+// Write kernel statistics to a stream
+std::ostream& operator<<(std::ostream& os, const KernelMetadata& md);
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Accumulate counters for a kernel launch.
+ */
+void KernelProfiling::log_launch(int num_threads)
+{
+    CELER_EXPECT(num_threads > 0);
+
+    // Increment launches by 1 and thread count by num_threads.
+    // We don't care in what order these values are written.
+    this->num_launches.fetch_add(1, std::memory_order_relaxed);
+    this->accum_threads.fetch_add(num_threads, std::memory_order_relaxed);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the kernel metadata for a given ID.
+ */
+auto KernelRegistry::at(key_type key) const -> const_reference
+{
+    CELER_EXPECT(key < this->size());
```

Comment:
This needs to be protected by the lock.  eg. the underlying kernels_ might be in the process of reallocation during its execution.  (and/or have some guarantee that those routines are never ever called when the insertion could happen)

Reply:
I'd like to keep it simple and say that this can't be accessed after insertion has completed. The "best" way I guess is to separate the "registration process" from the "registry", and move construct the registry from the registration after it's complete? That way the registration is thread safe and the accessing is not.

Alternatively, we could have a flag that fundamentally changes the class's state from "writeable" to "read-only" which is nasty.

Do you think the use pattern is clear enough that we can just leave the thread unsafety of the `at`/`size` methods as an "administrative control" rather than a "technical control"?